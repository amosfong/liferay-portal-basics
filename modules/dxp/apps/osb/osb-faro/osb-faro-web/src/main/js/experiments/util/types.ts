export type FormatYAxisFn = (metricUnit: string) => (value: number) => string;

export type GetFormattedMedianFn = (
	median: number,
	metric: MetricName
) => string;

export type GetMetricNameFn = (metric: MetricName) => string;

export type GetMetricUnitFn = (metric: MetricName) => string;

export type GetStatusColorFn = (status: Status) => string;

export type GetStatusNameFn = (status: Status) => string;

export type GetShortIntervals = (intervals: Array<Date>) => Array<Date>;

export type GetTicksFn = (maxValue: number) => Array<number>;

export type GetVariantLabels = (experiment: {
	bestVariant?: Variant;
	dxpVariantId: string;
	publishedDXPVariantId?: string;
	status: Status;
	winnerDXPVariantId?: string;
}) => Array<{status: string; value: string}>;

type MetricName =
	| 'BOUNCE_RATE'
	| 'CLICK_RATE'
	| 'MAX_SCROLL_DEPTH'
	| 'TIME_ON_PAGE';

type MergedVariant = {
	changes: number;
	confidenceInterval: Array<number>;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
	trafficSplit: number;
	uniqueVisitors: number;
};

export type MergedVariantsFn = (
	variants: Array<Variant>,
	variantMetrics: Array<VariantMetric>
) => Array<MergedVariant>;

type Status =
	| 'COMPLETED'
	| 'DRAFT'
	| 'FINISHED_NO_WINNER'
	| 'FINISHED_WINNER'
	| 'PAUSED'
	| 'RUNNING'
	| 'SCHEDULED'
	| 'TERMINATED';

type Variant = {
	changes: number;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	trafficSplit: number;
	uniqueVisitors: number;
};

type VariantMetric = {
	confidenceInterval: Array<number>;
	dxpVariantId: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
};
