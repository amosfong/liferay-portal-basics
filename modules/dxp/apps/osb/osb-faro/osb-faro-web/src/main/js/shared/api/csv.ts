import sendRequest from 'shared/util/request';
import {CSVType} from 'shared/components/download-report/utils';

export const fetchCSV = (url: string) => fetch(url);

export const fetchCount = ({
	groupId,
	type,
	...data
}: {
	assetId?: string;
	assetType?: string;
	channelId: string;
	fromDate?: string;
	groupId: string;
	individualId?: string;
	rangeKey?: string;
	toDate?: string;
	type: CSVType;
}) =>
	sendRequest({
		data,
		method: 'GET',
		path: `main/${groupId}/reports/export/csv/${type}/count`
	});
