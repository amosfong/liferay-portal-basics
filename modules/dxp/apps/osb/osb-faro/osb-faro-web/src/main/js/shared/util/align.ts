export const POSITIONS = {
	BottomCenter: 4,
	BottomLeft: 5,
	BottomRight: 3,
	LeftCenter: 6,
	RightCenter: 2,
	TopCenter: 0,
	TopLeft: 7,
	TopRight: 1
};

export type Position = typeof POSITIONS[keyof typeof POSITIONS];

export function align(element, alignElement, position, autoBestAlign = true) {
	let bestRegion;

	if (autoBestAlign) {
		const suggestion = suggestAlignBestRegion(
			element,
			alignElement,
			position
		);
		position = suggestion.position;
		bestRegion = suggestion.region;
	} else {
		bestRegion = getAlignRegion(element, alignElement, position);
	}

	const computedStyle = window.getComputedStyle(element, null);

	if (computedStyle.getPropertyValue('position') !== 'fixed') {
		bestRegion.top += window.scrollY;
		bestRegion.left += window.scrollX;

		let offsetParent = element;
		while ((offsetParent = offsetParent.offsetParent)) {
			bestRegion.top -= offsetParent.offsetTop;
			bestRegion.left -= offsetParent.offsetLeft;
		}
	}

	element.style.top = `${bestRegion.top}px`;
	element.style.left = `${bestRegion.left}px`;

	return position;
}

function getAlignRegion(element, alignElement, position) {
	const r1 = alignElement.getBoundingClientRect();
	const r2 = element.getBoundingClientRect();
	let top = 0;
	let left = 0;

	switch (position) {
		case POSITIONS.TopCenter:
		default:
			top = r1.top - r2.height;
			left = r1.left + r1.width / 2 - r2.width / 2;
			break;
		case POSITIONS.RightCenter:
			top = r1.top + r1.height / 2 - r2.height / 2;
			left = r1.left + r1.width;
			break;
		case POSITIONS.BottomCenter:
			top = r1.bottom;
			left = r1.left + r1.width / 2 - r2.width / 2;
			break;
		case POSITIONS.LeftCenter:
			top = r1.top + r1.height / 2 - r2.height / 2;
			left = r1.left - r2.width;
			break;
		case POSITIONS.TopRight:
			top = r1.top - r2.height;
			left = r1.right - r2.width;
			break;
		case POSITIONS.BottomRight:
			top = r1.bottom;
			left = r1.right - r2.width;
			break;
		case POSITIONS.BottomLeft:
			top = r1.bottom;
			left = r1.left;
			break;
		case POSITIONS.TopLeft:
			top = r1.top - r2.height;
			left = r1.left;
			break;
	}

	return {
		bottom: top + r2.height,
		height: r2.height,
		left,
		right: left + r2.width,
		top,
		width: r2.width
	};
}

function suggestAlignBestRegion(element, alignElement, position) {
	let bestArea = 0;
	let bestPosition = position;
	let bestRegion = getAlignRegion(element, alignElement, bestPosition);
	let tryPosition = bestPosition;
	let tryRegion = bestRegion;

	const viewportRegion = getViewportRegion();

	for (let i = 0; i < 8; ) {
		if (intersectRegion(viewportRegion, tryRegion)) {
			const visibleRegion = intersection(viewportRegion, tryRegion);

			const area = visibleRegion.width * visibleRegion.height;

			if (area > bestArea) {
				bestArea = area;
				bestRegion = tryRegion;
				bestPosition = tryPosition;
			}

			if (insideRegion(viewportRegion, tryRegion)) {
				break;
			}
		}
		tryPosition = (position + ++i) % 8;
		tryRegion = getAlignRegion(element, alignElement, tryPosition);
	}

	return {
		position: bestPosition,
		region: bestRegion
	};
}

function getViewportRegion() {
	const height = document.documentElement.offsetHeight;
	const width = document.documentElement.offsetWidth;

	return makeRegion(height, height, 0, width, 0, width);
}

function insideRegion(r1, r2) {
	return (
		r2.top >= r1.top &&
		r2.bottom <= r1.bottom &&
		r2.right <= r1.right &&
		r2.left >= r1.left
	);
}

function intersectRect(x0, y0, x1, y1, x2, y2, x3, y3) {
	return !(x2 > x1 || x3 < x0 || y2 > y1 || y3 < y0);
}

function intersectRegion(r1, r2) {
	return intersectRect(
		r1.top,
		r1.left,
		r1.bottom,
		r1.right,
		r2.top,
		r2.left,
		r2.bottom,
		r2.right
	);
}

function intersection(r1, r2) {
	if (!intersectRegion(r1, r2)) {
		return null;
	}
	const bottom = Math.min(r1.bottom, r2.bottom);
	const right = Math.min(r1.right, r2.right);
	const left = Math.max(r1.left, r2.left);
	const top = Math.max(r1.top, r2.top);

	return makeRegion(bottom, bottom - top, left, right, top, right - left);
}

function makeRegion(bottom, height, left, right, top, width) {
	return {
		bottom,
		height,
		left,
		right,
		top,
		width
	};
}
