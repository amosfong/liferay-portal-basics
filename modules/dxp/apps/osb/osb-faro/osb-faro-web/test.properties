##
## Test Suites
##

    modified.files.includes[relevant][osb-faro-web-playwright-rule]=**
    relevant.rule.names=osb-faro-web-playwright-rule

    test.batch.names[relevant][osb-faro-web-playwright-rule]=\
        playwright-js-tomcat90-mysql57

    #
    # Relevant
    #

    playwright.projects.includes[playwright-js-tomcat90-mysql57][relevant][osb-faro-web-playwright-rule]=\
        osb-faro-web

##
## Testray
##

    testray.main.component.name=Analytics Cloud