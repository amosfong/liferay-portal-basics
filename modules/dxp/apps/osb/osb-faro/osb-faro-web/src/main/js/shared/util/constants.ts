import {POSITIONS as ALIGN_POSITIONS} from './align';

export enum AcquisitionTypes {
	Channel = 'CHANNEL',
	Referrer = 'REFERRER',
	SourceMedium = 'SOURCE_MEDIUM'
}

export enum ActivityActions {
	Comments = 3,
	Downloads = 0,
	Previews = 4,
	Submissions = 1,
	Visits = 2
}

export enum Applications {
	Contacts = 'contacts',
	Main = 'main'
}

export enum AssetNames {
	BlogViewed = 'blogViewed',
	CommentPosted = 'commentPosted',
	DocumentDownloaded = 'documentDownloaded',
	DocumentPreviewed = 'documentPreviewed',
	FormSubmitted = 'formSubmitted',
	FormViewed = 'formViewed',
	PageViewed = 'pageViewed',
	WebContentViewed = 'webContentViewed'
}

export enum AssetTypes {
	Asset = 'Asset',
	Blog = 'Blog',
	Document = 'Document',
	Form = 'Form',
	Journal = 'Journal',
	WebContent = 'WebContent',
	WebPage = 'Page'
}

export enum ChannelPermissionTypes {
	AllUsers = 0,
	SelectUsers = 1
}

export enum CompositionTypes {
	AccountInterests = 'accountInterests',
	Acquisitions = 'acquisitions',
	IndividualInterests = 'individualInterests',
	SearchTerms = 'searchTerms',
	SegmentInterests = 'individualSegmentInterests',
	SiteInterests = 'siteInterests'
}

export enum ConjunctionKey {
	And = 'and',
	Or = 'or'
}

export enum CredentialTypes {
	OAuth1 = 'OAuth 1 Authentication',
	OAuth2 = 'OAuth 2 Authentication',
	Token = 'Token Authentication'
}

export enum DataSourceDisplayStatuses {
	Active = 'ACTIVE',
	Configuring = 'CONFIGURING',
	DeleteError = 'DELETE_ERROR',
	InDeletion = 'IN_DELETION',
	Inactive = 'INACTIVE'
}

export enum DataSourceProgressStatuses {
	Completed = 'COMPLETED',
	Failed = 'FAILED',
	InProgress = 'IN_PROGRESS',
	Started = 'STARTED'
}

export enum DataSourceStates {
	ActionNeeded = 'ACTION_NEEDED',
	AnalyticsClientConfigurationFailure = 'ANALYTICS_CLIENT_CONFIGURATION_FAILURE',
	CredentialsInvalid = 'CREDENTIALS_INVALID',
	CredentialsValid = 'CREDENTIALS_VALID',
	Disconnected = 'DISCONNECTED',
	InProgressDeleting = 'IN_PROGRESS_DELETING',
	LiferayVersionInvalid = 'LIFERAY_VERSION_INVALID',
	UndefinedError = 'UNDEFINED_ERROR',
	Ready = 'READY',
	UrlInvalid = 'URL_INVALID',
	Unconfigured = 'UNCONFIGURED'
}
export enum DataSourceStatuses {
	Active = 'ACTIVE',
	Inactive = 'INACTIVE'
}

export enum DataSourceTypes {
	Csv = 'CSV',
	Liferay = 'LIFERAY',
	Salesforce = 'SALESFORCE'
}

export enum EntityTypes {
	Account = 0,
	AccountsSegment = 3,
	Asset = 5,
	DataSource = 1,
	Individual = 2,
	IndividualsSegment = 4,
	Page = 6
}

export enum ExpirationPeriod {
	In30Days = '2592000',
	In6Months = '15778800',
	In1Year = '31557600',
	Indefinite = '3155760000'
}

export enum FaroEnv {
	Local = 'local',
	Production = 'prd',
	Staging = 'stg'
}

export enum FieldContexts {
	Custom = 'custom',
	Demographics = 'demographics',
	Interests = 'interests',
	Organization = 'organization'
}

export enum FieldOwnerTypes {
	Account = 'account',
	Individual = 'individual',
	Organization = 'organization'
}

export enum FieldTypes {
	Boolean = 'Boolean',
	Date = 'Date',
	Number = 'Number',
	String = 'Text'
}

export enum GDPRRequestStatuses {
	Completed = 'COMPLETED',
	Error = 'ERROR',
	Expired = 'EXPIRED',
	Pending = 'PENDING',
	Running = 'RUNNING'
}

export enum GDPRRequestTypes {
	Access = 'ACCESS',
	Delete = 'DELETE',
	Suppress = 'SUPPRESS',
	Unsuppress = 'UNSUPPRESS'
}

export enum JobRunDataPeriods {
	Last7Days = 'LAST_7_DAYS',
	Last30Days = 'LAST_30_DAYS',
	Last180Days = 'LAST_180_DAYS',
	Last365Days = 'LAST_365_DAYS'
}

export enum JobRunFrequencies {
	Every7Days = 'EVERY_7_DAYS',
	Every14Days = 'EVERY_14_DAYS',
	Every30Days = 'EVERY_30_DAYS',
	Manual = 'MANUAL'
}

export enum JobRunStatuses {
	Completed = 'COMPLETED',
	Failed = 'FAILED',
	Published = 'PUBLISHED',
	Running = 'RUNNING'
}

export enum JobStatuses {
	Failed = 'FAILED',
	Pending = 'PENDING',
	Ready = 'READY',
	Running = 'RUNNING',
	Scheduled = 'SCHEDULED'
}

export enum JobTypes {
	ItemSimilarity = 'CONTENT_RECOMMENDATION_ITEM_SIMILARITY'
}

export enum LanguageIds {
	English = 'en_US',
	Japanese = 'ja_JP',
	Portuguese = 'pt_BR',
	Spanish = 'es_ES'
}

export enum OrderByDirections {
	Ascending = 'ASC',
	Descending = 'DESC'
}

export enum PreferencesScopes {
	User = 'user',
	Group = 'group'
}

export enum ProjectStates {
	Activating = 'ACTIVATING',
	AutoRedeployFailed = 'AUTO_REDEPLOY_FAILED',
	Deactivated = 'DEACTIVATED',
	Maintenance = 'MAINTENANCE',
	NotReady = 'NOT READY',
	Ready = 'READY',
	Scheduled = 'SCHEDULED',
	Unavailable = 'UNAVAILABLE',
	Unconfigured = 'UNCONFIGURED'
}

export enum RangeKeyTimeRanges {
	CustomRange = 'CUSTOM',
	Last180Days = '180',
	Last24Hours = '0',
	Last28Days = '28',
	Last30Days = '30',
	Last7Days = '7',
	Last90Days = '90',
	LastYear = '365',
	Yesterday = '1'
}

export enum SegmentStates {
	Disabled = 'DISABLED',
	InProgress = 'IN_PROGRESS',
	Ready = 'READY'
}

export enum SegmentTypes {
	Dynamic = 'DYNAMIC',
	Static = 'STATIC'
}

export enum SessionEntityTypes {
	Account = 'ACCOUNT',
	Individual = 'INDIVIDUAL'
}

export enum Sizes {
	Small = 'sm',
	Medium = 'md',
	Large = 'lg',
	XLarge = 'xl',
	XXLarge = 'xxl',
	XXXLarge = 'xxxl'
}

export enum SubscriptionStatuses {
	Approaching = 1,
	Ok = 0,
	Over = 2
}

export enum TimeIntervals {
	Day = 'day',
	Month = 'month',
	Quarter = 'quarter',
	Week = 'week',
	Year = 'year'
}

export enum TimeSpans {
	AllTime = 'ever',
	LastYear = 'lastYear',
	Last24Hours = 'last24Hours',
	Last7Days = 'last7Days',
	Last28Days = 'last28Days',
	Last30Days = 'last30Days',
	Last90Days = 'last90Days',
	Today = 'today',
	Yesterday = 'yesterday'
}

export enum UserRoleNames {
	Administrator = 'Site Administrator',
	Member = 'Site Member',
	Owner = 'Site Owner'
}

export enum UserStatuses {
	Approved = 0,
	Pending = 1,
	Requested = 2
}

export const ALIGNMENTS_MAP = {
	bottom: ALIGN_POSITIONS.BottomCenter,
	'bottom-left': ALIGN_POSITIONS.BottomLeft,
	'bottom-right': ALIGN_POSITIONS.BottomRight,
	left: ALIGN_POSITIONS.LeftCenter,
	right: ALIGN_POSITIONS.RightCenter,
	top: ALIGN_POSITIONS.TopCenter,
	'top-left': ALIGN_POSITIONS.TopLeft,
	'top-right': ALIGN_POSITIONS.TopRight
};

export const ASSET_METRICS = [
	{
		key: 'abandonmentsMetric',
		selectTitle: `${Liferay.Language.get(
			'form-abandonment'
		)} (${Liferay.Language.get('ratio').toLowerCase()})`,
		title: Liferay.Language.get('form-abandonment'),
		type: 'percentage'
	},
	{
		key: 'clicksMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-clicks'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('clicks'),
		type: 'number'
	},
	{
		key: 'completionTimeMetric',
		selectTitle: `${Liferay.Language.get(
			'form-completion-time'
		)} (${Liferay.Language.get('average').toLowerCase()})`,
		title: Liferay.Language.get('completion-time'),
		type: 'time'
	},
	{
		key: 'downloadsMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-downloads'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('downloads'),
		type: 'number'
	},
	{
		key: 'readingTimeMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-interaction-time'
		)} (${Liferay.Language.get('average').toLowerCase()})`,
		title: Liferay.Language.get('interaction-time'),
		type: 'time'
	},
	{
		key: 'submissionsMetric',
		selectTitle: `${Liferay.Language.get(
			'form-submissions'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('form-submissions'),
		type: 'number'
	},
	{
		key: 'viewsMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-views'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('views'),
		type: 'number'
	}
];

export const DATA_RETENTION_PERIOD_KEY = 'data-retention-period';

export const DEVELOPER_MODE = FARO_DEV_MODE;

// LRAC-11571 Disable temporarily Accounts

export const ENABLE_ACCOUNTS = false;

export const ENABLE_ADD_TRIAL_WORKSPACE =
	FARO_ENV === FaroEnv.Local || FARO_ENV === FaroEnv.Staging;

// LRAC - 11652 Hide Displayed Asset Card on the Page report

export const ENABLE_ASSET_CARD = false;

// LRAC-13649 Hide the keywords blocklist screen from Settings>Definitions

export const ENABLE_BLOCKLIST_KEYWORDS = false;

// LRAC-11651 Disable temporarily CSV File

export const ENABLE_CSVFILE = false;

// LRAC-13389 Disable temporarily Delete Property and Delete Data Source buttons

export const ENABLE_DELETE_DATA_SOURCE_BUTTON = false;

export const ENABLE_DELETE_PROPERTY_BUTTON = false;

// LRAC-11650 Hide Form Abandonment Card

export const ENABLE_FORM_ABANDONMENT = false;

// LRAC-10908 Hide global filters

export const ENABLE_GLOBAL_FILTER = false;

// LRAC-13781 [BUG] The Last Seen column only displays the creation date of the token

export const ENABLE_LAST_ACCESS_DATE = false;

// LRAC-11571 Disable temporarily Salesforce

export const ENABLE_SALESFORCE = false;

export const EXPIRATION_DATE_LABELS = {
	[ExpirationPeriod.In30Days]: Liferay.Language.get('30-days'),
	[ExpirationPeriod.In6Months]: Liferay.Language.get('6-months'),
	[ExpirationPeriod.In1Year]: Liferay.Language.get('1-year'),
	[ExpirationPeriod.Indefinite]: Liferay.Language.get('indefinite')
};

export const LANGUAGES = [
	{
		id: LanguageIds.English,
		label: Liferay.Language.get('english')
	},
	{
		id: LanguageIds.Japanese,
		label: Liferay.Language.get('japanese')
	},
	{
		id: LanguageIds.Portuguese,
		label: Liferay.Language.get('portuguese')
	},
	{
		id: LanguageIds.Spanish,
		label: Liferay.Language.get('spanish')
	}
];

export const ONE_DAY = '86400000';

export const ONE_MONTH = '2592000000';

export const POSITIONS = [
	'top',
	'top',
	'right',
	'bottom',
	'bottom',
	'bottom',
	'left',
	'top'
];

export const SEVEN_MONTHS = '18144000000';

export const THIRTEEN_MONTHS = '33696000000';

export const TIME_RANGE_LABELS = {
	[RangeKeyTimeRanges.Last180Days]: Liferay.Language.get('last-180-days'),
	[RangeKeyTimeRanges.Last24Hours]: Liferay.Language.get('last-24-hours'),
	[RangeKeyTimeRanges.Last28Days]: Liferay.Language.get('last-28-days'),
	[RangeKeyTimeRanges.Last30Days]: Liferay.Language.get('last-30-days'),
	[RangeKeyTimeRanges.Last7Days]: Liferay.Language.get('last-seven-days'),
	[RangeKeyTimeRanges.Last90Days]: Liferay.Language.get('last-90-days'),
	[RangeKeyTimeRanges.LastYear]: Liferay.Language.get('last-year'),
	[RangeKeyTimeRanges.Yesterday]: Liferay.Language.get('yesterday')
};

export const TWO_DAYS = '172800000';

const Constants: Window['faroConstants'] = window.faroConstants;

export default Constants;
