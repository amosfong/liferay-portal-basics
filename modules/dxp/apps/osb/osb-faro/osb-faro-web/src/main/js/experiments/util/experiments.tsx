import {CHART_COLORS} from 'shared/util/charts';
import {CONTROL_COLOR} from './constants';
import {
	FormatYAxisFn,
	GetFormattedMedianFn,
	GetMetricNameFn,
	GetMetricUnitFn,
	GetShortIntervals,
	GetStatusColorFn,
	GetStatusNameFn,
	GetTicksFn,
	GetVariantLabels,
	MergedVariantsFn
} from './types';
import {round} from 'lodash';
import {toRounded, toThousands, toThousandsBase} from 'shared/util/numbers';

const METRICS_NAMES = new Map([
	['BOUNCE_RATE', Liferay.Language.get('bounce-rate')],
	['CLICK_RATE', Liferay.Language.get('click-through-rate')],
	['MAX_SCROLL_DEPTH', Liferay.Language.get('max-scroll-depth')],
	['TIME_ON_PAGE', Liferay.Language.get('view-duration')]
]);

const METRICS_UNITS = new Map([
	['BOUNCE_RATE', '%'],
	['CLICK_RATE', '%'],
	['MAX_SCROLL_DEPTH', '%'],
	['TIME_ON_PAGE', 's']
]);

const STATUS_COLORS = new Map([
	['COMPLETED', 'success'],
	['DRAFT', 'secondary'],
	['FINISHED_NO_WINNER', 'secondary'],
	['FINISHED_WINNER', 'success'],
	['PAUSED', 'secondary'],
	['RUNNING', 'info'],
	['SCHEDULED', 'warning'],
	['TERMINATED', 'danger']
]);

const STATUS_NAMES = new Map([
	['COMPLETED', Liferay.Language.get('complete')],
	['DRAFT', Liferay.Language.get('draft')],
	['FINISHED_NO_WINNER', Liferay.Language.get('no-winner')],
	['FINISHED_WINNER', Liferay.Language.get('winner-declared')],
	['PAUSED', Liferay.Language.get('paused')],
	['RUNNING', Liferay.Language.get('running')],
	['SCHEDULED', Liferay.Language.get('scheduled')],
	['TERMINATED', Liferay.Language.get('terminated')]
]);

const getExperimentLink = ({action, id, pageURL}) => {
	const experimentLink = `${pageURL}?segmentsExperimentKey=${id}`;

	if (action) {
		return `${experimentLink}&segmentsExperimentAction=${action}`;
	}

	return experimentLink;
};

export const formatYAxis: FormatYAxisFn = metricUnit => value => {
	if (value % 1 === 0) {
		return `${value}${metricUnit}`;
	}

	return `${value.toFixed(1)}${metricUnit}`;
};

export const getFormattedMedian: GetFormattedMedianFn = (median, metric) => {
	const precision = metric === 'CLICK_RATE' ? 3 : 2;

	return toRounded(median, precision);
};

export const getFormattedMedianLabel = metric =>
	metric === 'CLICK_RATE'
		? `${Liferay.Language.get('median')} ${getMetricName(metric)}`
		: `${getMetricName(metric)} ${Liferay.Language.get('median')}`;

export const getFormattedProbabilityToWin = value => {
	if (value < 0.1) {
		value = '< 0.1';
	} else if (value > 99.9) {
		value = '> 99.9';
	} else {
		value = toRounded(value);
	}

	return value;
};

export const getMetricName: GetMetricNameFn = metric =>
	METRICS_NAMES.get(metric);

export const getMetricUnit: GetMetricUnitFn = metric =>
	METRICS_UNITS.get(metric);

export const getStatusColor: GetStatusColorFn = status =>
	STATUS_COLORS.get(status);

export const getStatusName: GetStatusNameFn = status =>
	STATUS_NAMES.get(status).toUpperCase();

export const mergedVariants: MergedVariantsFn = (variants, variantMetrics) =>
	variants.map(variant => ({
		...variant,
		...variantMetrics.find(
			({dxpVariantId}) => variant.dxpVariantId === dxpVariantId
		)
	}));

export const getActions = (
	status: string,
	{id, onDelete, pageURL, publishable} = null
) => {
	const deleteButton = {
		displayType: 'secondary',
		label: Liferay.Language.get('delete'),
		onClick: onDelete
	};

	switch (status) {
		case 'COMPLETED': {
			return [deleteButton];
		}
		case 'DRAFT': {
			return [
				{
					displayType: 'primary',
					label: Liferay.Language.get('review'),
					redirectURL: getExperimentLink({
						action: 'reviewAndRun',
						id,
						pageURL
					})
				},
				{
					displayType: 'secondary',
					label: Liferay.Language.get('delete'),
					redirectURL: getExperimentLink({
						action: 'delete',
						id,
						pageURL
					})
				}
			];
		}
		case 'FINISHED_NO_WINNER':
		case 'FINISHED_WINNER': {
			return [
				{
					displayType: 'primary',
					label: Liferay.Language.get('publish'),
					redirectURL: getExperimentLink({
						action: 'publish',
						id,
						pageURL
					})
				},
				{
					displayType: 'secondary',
					label: Liferay.Language.get('delete'),
					redirectURL: getExperimentLink({
						action: 'delete',
						id,
						pageURL
					})
				}
			];
		}
		case 'TERMINATED': {
			if (publishable) {
				return [
					{
						displayType: 'primary',
						label: Liferay.Language.get('publish'),
						redirectURL: getExperimentLink({
							action: 'publish',
							id,
							pageURL
						})
					},
					{
						displayType: 'secondary',
						label: Liferay.Language.get('delete'),
						redirectURL: getExperimentLink({
							action: 'delete',
							id,
							pageURL
						})
					}
				];
			}

			return [deleteButton];
		}
		case 'RUNNING': {
			return [
				{
					displayType: 'secondary',
					label: Liferay.Language.get('terminate'),
					redirectURL: getExperimentLink({
						action: 'terminate',
						id,
						pageURL
					})
				}
			];
		}
		default: {
			return [];
		}
	}
};

export const getBestVariant = ({
	dxpVariants,
	goal,
	metrics: {variantMetrics}
}) => {
	if (
		!dxpVariants ||
		variantMetrics.every(({median}) => median === variantMetrics[0].median)
	) {
		return null;
	}

	if (goal?.metric === 'BOUNCE_RATE') {
		return mergedVariants(
			dxpVariants,
			variantMetrics
		).reduce((prev, current) =>
			prev.median < current.median ? prev : current
		);
	}

	return mergedVariants(dxpVariants, variantMetrics).reduce((prev, current) =>
		prev.median > current.median ? prev : current
	);
};

export const getVariantLabels: GetVariantLabels = ({
	bestVariant,
	dxpVariantId,
	publishedDXPVariantId,
	status,
	winnerDXPVariantId
}) => {
	const labels = [];

	if (status === 'RUNNING' && bestVariant?.dxpVariantId === dxpVariantId) {
		labels.push({
			status: 'success',
			value: Liferay.Language.get('current-best')
		});
	}

	if (
		winnerDXPVariantId === dxpVariantId &&
		(status === 'COMPLETED' || status === 'FINISHED_WINNER')
	) {
		labels.push({
			status: 'success',
			value: Liferay.Language.get('winner')
		});
	}

	if (publishedDXPVariantId === dxpVariantId) {
		labels.push({
			status: 'info',
			value: Liferay.Language.get('published')
		});
	}

	return labels;
};

export const getTicks: GetTicksFn = maxValue => {
	const arr = [];
	let interval = 1;
	const step = Math.round(maxValue / 8);

	while (interval <= maxValue) {
		arr.push(interval);

		interval = interval + step;
	}

	return [...arr];
};

export const getShortIntervals: GetShortIntervals = intervals =>
	getTicks(intervals.length).map(tick => intervals[tick - 1]);

export const toThousandsABTesting = number => {
	if (number > 1e4) {
		return toThousandsBase(number, factor =>
			Math.trunc(round(number * factor, 2))
		);
	}

	return toThousands(number);
};

export const getLegendData = dxpVariants => {
	const COLORS = [...CHART_COLORS];

	return dxpVariants.map(({control, dxpVariantId, dxpVariantName}) => ({
		color: control ? CONTROL_COLOR : COLORS.shift(),
		id: dxpVariantId,
		name: dxpVariantName
	}));
};

export const getMedianGraphData = ({dxpVariants, metricUnit}) => {
	const COLORS = [...CHART_COLORS];

	const type = metricUnit === '%' ? 'percentage' : 'number';

	const formatter =
		metricUnit === '%' ? value => value : value => `${value}s`;

	const items = dxpVariants.map(({confidenceInterval, control, median}) => ({
		intervals: [
			{
				end: confidenceInterval[1],
				start: confidenceInterval[0]
			}
		],
		progress: [
			{
				color: control ? CONTROL_COLOR : COLORS.shift(),
				value: median
			}
		]
	}));

	const maxValue = Math.max(
		...dxpVariants.map(({confidenceInterval}) => confidenceInterval[1])
	);

	return {
		disableScroll: true,
		empty: maxValue === 0,
		formatSpacement: false,
		grid: {
			formatter,
			maxValue,
			minValue: 0,
			precision: 2,
			show: true,
			type
		},
		items
	};
};
