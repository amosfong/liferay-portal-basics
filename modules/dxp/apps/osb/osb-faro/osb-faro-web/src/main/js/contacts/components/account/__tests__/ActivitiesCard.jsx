import * as API from 'shared/api';
import * as data from 'test/data';
import ActivitiesCard from '../ActivitiesCard';
import React from 'react';
import {Account} from 'shared/util/records';
import {render} from '@testing-library/react';
import {StaticRouter} from 'react-router';
import {waitForLoadingToBeRemoved} from 'test/helpers';

jest.unmock('react-dom');

describe('ActivitiesCard', () => {
	it('should render', async () => {
		const {container} = render(
			<StaticRouter>
				<ActivitiesCard
					account={data.getImmutableMock(
						Account,
						data.mockAccount,
						'test'
					)}
					groupId='23'
				/>
			</StaticRouter>
		);

		jest.runAllTimers();

		await waitForLoadingToBeRemoved(container);

		expect(container).toMatchSnapshot();
	});

	it('should render w/ loading', () => {
		const {container} = render(
			<StaticRouter>
				<ActivitiesCard
					account={data.getImmutableMock(
						Account,
						data.mockAccount,
						'test'
					)}
					groupId='23'
				/>
			</StaticRouter>
		);

		expect(container.querySelector('.loading-root')).toBeTruthy();
	});

	it('should render w/ ErrorDisplay', async () => {
		API.activities.fetchHistory.mockReturnValueOnce(Promise.reject({}));

		const {container, getByText} = render(
			<StaticRouter>
				<ActivitiesCard
					account={data.getImmutableMock(
						Account,
						data.mockAccount,
						'test'
					)}
					groupId='23'
				/>
			</StaticRouter>
		);

		jest.runAllTimers();

		await waitForLoadingToBeRemoved(container);

		expect(getByText('An unexpected error occurred.')).toBeTruthy();
	});
});
