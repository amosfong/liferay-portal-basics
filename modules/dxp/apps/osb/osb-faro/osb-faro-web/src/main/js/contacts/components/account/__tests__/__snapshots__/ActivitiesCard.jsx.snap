// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivitiesCard should render 1`] = `
<div>
  <div
    class="card card-root account-activities-card-root"
  >
    <div
      class="card-header"
    >
      <div
        class="card-title"
      >
        Account Activities
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="change-legend-root"
      >
        <div
          class="legend-item activities"
        >
          <div
            class="legend-header"
          >
            <span
              class="legend-color"
              style="background-color: rgb(75, 155, 255);"
            />
            <span
              class="title"
            >
              Total Activity Count: 123
            </span>
          </div>
          <div
            class="secondary-info"
          >
            30 Day Change
            <span
              class="change increase"
            >
              <svg
                class="lexicon-icon lexicon-icon-caret-top icon-root"
                role="presentation"
              >
                <use
                  href="#caret-top"
                />
              </svg>
              128.0%
            </span>
          </div>
        </div>
      </div>
      <div
        class="composed-chart-with-empty-state"
      >
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 340px; min-width: 0;"
        />
      </div>
    </div>
    <div
      class="card-footer"
    >
      <a
        class="button-root btn btn-outline-borderless btn-sm btn-outline-secondary"
        href="/workspace/23/contacts/accounts/test/activities"
      >
        View All Activities
        <svg
          class="lexicon-icon lexicon-icon-angle-right-small icon-root ml-2"
          role="presentation"
        >
          <use
            href="#angle-right-small"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
